addon/services/toast.ts

```diff
@@ -85,123 +93,179 @@ export default class Toast extends Service {
-    const container: Element = this._buildContainer();
-    const toast: Element = this._buildElement('div', ['toast', 'toast--hidden', `toast-${type}`]);
-    let toastChildren: Element[] = [this._buildElement('div', ['toast-message'], message)];
+  private _createCounterAnimation(element: HTMLElement, opts: ToastOptions): Animation {
+    const duration = typeof opts.timeout === 'number' ? opts.timeout : DEFAULT_TOAST_TIMEOUT;
+    const counterDownKeyframes = new KeyframeEffect(element, [{ width: '100%' }, { width: '0' }], {
+      duration: duration,
+      fill: 'forwards'
+    });
 
-    if (title && !isEmpty(title)) {
-      toastChildren.splice(0, 0, this._buildElement('div', ['toast-title'], title));
-    }
+    return new Animation(counterDownKeyframes, document.timeline);
+  }
 
-    if (opts.closeButton) {
-      toastChildren.push(this._buildCloseButton());
-    }
+  private _createDestroyAnimation(element: HTMLElement): Animation {
+    const destroyDownKeyframes = new KeyframeEffect(
+      element,
+      [
+        { opacity: 1, transform: 'translate(0, 0)' },
+        { opacity: 0, transform: 'translate(0, -50%)' }
+      ],
+      {
+        duration: 300,
+        fill: 'forwards'
+      }
+    );
 
-    this._setupToastEvents(toast, opts);
+    return new Animation(destroyDownKeyframes, document.timeline);
+  }
 
-    toastChildren.forEach((x) => toast.append(x));
+  private _playCounterAnimation(event: Event): void {
+    this._toasts.get(<HTMLElement>event.target)?.play();
+  }
 
-    if (container.querySelector('.toast')) {
-      container.insertBefore(toast, container.querySelector('.toast'));
+  private _pauseCounterAnimation(event: Event): void {
+    this._toasts.get(<HTMLElement>event.target)?.pause();
+  }
+
+  private _buildToast(type: ToastType, message: string, title: string | undefined, opts: ToastOptions): void {
+    const container: HTMLElement = this._buildContainer();
+    const toast: HTMLElement = this._buildElement('div', ['toast', 'toast--hidden', `toast--${type}`]);
+    const mainContainer: HTMLElement = this._buildElement('div', ['main-container']);
+
+    const counter = this._buildElement('div', ['counter']);
+    const counterDownAnimation = this._createCounterAnimation(counter, opts);
+
+    if (opts?.badge) {
+      let badgeContent: string = '';
+      if (opts.badge?.image) {
+        badgeContent = `<img src=${opts.badge.image} alt="" />`;
+      } else if (opts.badge?.icon) {
+        badgeContent = `<i class="${opts.badge.icon}"></i>`;
+      } else if (opts.badge?.text) {
+        badgeContent = `<span class="upf-badge__text">${opts.badge.text}</span>`;
+      }
+
+      const badgeContainer: HTMLElement = this._buildElement(
+        'div',
+        ['upf-badge', 'upf-badge--shape-round', 'upf-badge--size-md'],
+        badgeContent
+      );
+      mainContainer.append(badgeContainer);
```

Comment:
Maybe move this logic into another private method for better readability ? 
You might be able to do the same for the code below as well ?